packages/rOpenScPCA/R/sweep-clusters.R

```diff
@@ -0,0 +1,127 @@
+#'   if a SingleCellExperiment or Seurat object is provided. If not provided, the SingleCellExperiment
+#'   object name will default to "PCA" and the Seurat object name will default to "pca".
+#'
+#' @return A list of data frames from performing clustering across all parameter combinations.
+#'   Columns include `cluster_set` (identifier column for results from a single clustering run),
+#'   `cell_id`, and `cluster`. Additional columns represent algorithm parameters and include at least:
+#'   `algorithm`, `weighting`, and `nn`. Louvain and leiden clustering will also include `resolution`,
+#'   and leiden clustering will further include `objective_function`.
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # perform louvain clustering with jaccard weighting (defaults),
+#' # varying the nearest neighobor parameter.
+#' cluster_df <- sweep_clusters(sce_object, nn = c(10, 15, 20, 25))
+#'
+#' # perform louvain clustering, with jaccard and rank weighting, and
+#' # varying the nearest neighbor and resolution parameters.
+#' cluster_df <- sweep_clusters(
+#'   sce_object,
+#'   algorithm = "louvain",
+#'   weighting = c("jaccard", "rank"),
+#'   nn = c(10, 15, 20, 25),
+#'   resolution = c(0.5, 1)
+#' )
+#'
+#' # perform walktrap and louvain clustering with jaccard weighting, and
+#' # varying the nearest neighbors for both algorithms, and resolution for louvain.
+#' cluster_df <- sweep_clusters(
+#'   sce_object,
+#'   algorithm = c("walktrap", "louvain"),
+#'   weighting = "jaccard",
+#'   nn = c(10, 15, 20, 25),
+#'   resolution = c(0.5, 1)
+#' )
+#' }
+sweep_clusters <- function(
+    x,
+    algorithm = "louvain",
+    weighting = "jaccard",
+    nn = 10,
+    resolution = 1, # louvain or leiden
+    objective_function = "CPM", # leiden only
+    threads = 1,
+    seed = NULL,
+    pc_name = NULL) {
+  # Ensure input is a matrix for slightly faster processing later
+  if (any(class(x) %in% c("matrix", "Matrix"))) {
+    stopifnot(
+      "The matrix must have row names representing cell ids, e.g. barcodes." = is.character(rownames(x))
+    )
+  } else if (is(x, "SingleCellExperiment") || is(x, "Seurat")) {
+    x <- extract_pc_matrix(x, pc_name = pc_name)
+  } else {
+    stop("The first argument should be one of: a SingleCellExperiment object, a Seurat object, or a matrix with row names.")
+  }
+
+  # Collect all specific inputs into a single list
+  sweep_params <- tidyr::expand_grid(
+    algorithm = unique(algorithm),
+    weighting = unique(weighting),
+    nn = unique(nn),
+    resolution = unique(resolution),
+    objective_function = unique(objective_function)
+  ) |>
+    # set unused parameters for the given algorithm to their defaults, since we can't provide NA
+    # to match.arg in calculate_clusters
+    dplyr::mutate(
+      resolution = ifelse(algorithm %in% c("louvain", "leiden"), resolution, 1),
+      objective_function = ifelse(algorithm == "leiden", objective_function, "CPM")
```

Comment:
Okay, I looked at the function more closely, and it does appear that these unused parameters are fully not used and are not included in the table, so we should be fine.